sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/valueentity/ReflectiveValueEntityRouter.scala

```diff
@@ -54,4 +56,26 @@ class ReflectiveValueEntityRouter[S, E <: ValueEntity[S]](
       .invoke(entity, invocationContext)
       .asInstanceOf[ValueEntity.Effect[_]]
   }
+
+  private def _extractAndSetCurrentState(state: S): Unit = {
+    val entityStateType: Class[S] =
+      this.entity.getClass.getGenericSuperclass
+        .asInstanceOf[ParameterizedType]
+        .getActualTypeArguments
+        .head
+        .asInstanceOf[Class[S]]
+
+    // the state: S received can either be of the entity "state" type (if coming from emptyState/memory)
+    // or PB Any type (if coming from the proxy)
+    state match {
+      case s if s == null || state.getClass == entityStateType =>
+        // note that we set the state even if null, this is needed in order to
+        // be able to call currentState() later
+        entity._internalSetCurrentState(s)
+      case s =>
+        val deserializedState =
+          JsonSupport.decodeJson(entityStateType, ScalaPbAny.toJavaProto(s.asInstanceOf[ScalaPbAny]))
+        entity._internalSetCurrentState(deserializedState)
+    }
+  }
```

Comment:
we now have this code in 3 different places (ES, VE and views router) but I'm not sure we want to extract it to a common place tbh... might have a look again later or let me know if you have an opinion.

Reply:
Yep, we can extract the transformation to some utility class, but also fine to keep it as such for accelerate the release. We can move the code later.